tests/src/JIT/Performance/CodeQuality/Span/Indexer.cs

```diff
@@ -0,0 +1,925 @@
+                return result;
+            },
+            "KnownSizeCtor({0})", length);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static byte TestKnownSizeCtor(byte[] a, int innerIterationCount)
+        {
+            Span<byte> data = new Span<byte>(a, 0, 1024);
+            byte x = 0;
+
+            for (int i = 0; i < innerIterationCount; i++)
+            {
+                x = 0;
+                for (var idx = 0; idx < data.Length; idx++)
+                {
+                    x ^= data[idx];
+                }
+            }
+
+            return x;
+        }
+
+        [Benchmark(InnerIterationCount = Iterations)]
+        [InlineData(DefaultLength)]
+        [Category("Span known size bounds check elimination")]
+        public static void KnownSizeCtor2(int length)
+        {
+            if (length < 1024)
+            {
+                throw new Exception("test requires at least 1024 byte length");
+            }
+
+            byte[] a = GetData(length);
+
+            Invoke((int innerIterationCount) =>
+            {
+                byte result = TestKnownSizeCtor2(a, innerIterationCount);
+                return result;
+            },
+            "KnownSizeCtor2({0})", length);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static byte TestKnownSizeCtor2(byte[] a, int innerIterationCount)
+        {
+            Span<byte> data1 = new Span<byte>(a, 0, 512);
+            Span<byte> data2 = new Span<byte>(a, 512, 512);
+            byte x = 0;
+
+            for (int i = 0; i < innerIterationCount; i++)
+            {
+                x = 0;
+                for (var idx = 0; idx < data1.Length; idx++)
+                {
+                    x ^= data1[idx];
+                }
+                for (var idx = 0; idx < data2.Length; idx++)
+                {
+                    x ^= data2[idx];
+                }
+            }
+
+            return x;
+        }
+
+        [Benchmark(InnerIterationCount = Iterations)]
+        [InlineData(DefaultLength)]
+        [Category("Same index in-loop redundant bounds check elimination")]
+        public static void SameIndex1(int length)
+        {
+            byte[] a = GetData(length);
+
+            Invoke((int innerIterationCount) =>
+            {
+                Span<byte> s = new Span<byte>(a);
+                byte result = 0;
+                for (int i = 0; i < innerIterationCount; ++i)
+                {
+                    result = TestSameIndex1(s);
+                }
+                return result;
+            },
+            "SameIndex1({0})", length);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static byte TestSameIndex1(Span<byte> data)
+        {
+            byte x = 0;
+            byte y = 0;
+
+            for (var idx = 0; idx < data.Length; idx++)
+            {
+                x ^= data[idx];
+                y ^= data[idx];
+            }
```

Comment:
In `SameIndex1`, `SameIndex2`, and `CoveredIndex1`, I think you're writing the tests to ensure that we can remove a second bounds check based on a prior bounds check succeeding, in straight-line code.  But since you put the straight-line code in a loop and the indexer is the IV ranging from 0 to data.Length, once we handle the IV cases we'll be removing both range checks from the loop and no longer testing what you intended (if I'm following correctly).  Might want to make the loop bound an argument to the inner method or something.

It looks like maybe the same thing is happening in `WriteViaIndexer2`, but there you used 1 for the initial value of the IV, not sure if you were counting on that ruling out the IV-based removal or not.